Chatkit/Auth/TokenProvider.swift

```diff
@@ -0,0 +1,137 @@
+                ),
+                retryAttempts: 3,
+                retryDelayMs: 1000
+            )
+        )
+    }
+
+    internal static func testTokenHTTPCallProvider(instanceLocator: String, userId: String) -> TokenProvider {
+        // TODO: Implement:
+        // extract host and instanceId from instanceLocator
+        let host = "us1.pusherplatform.io"
+        let instanceId = "UNIMPLEMENTED"
+        let path = "/services/chatkit_token_provider/v1/\(instanceId)/token"
+
+        return HTTPCallTokenProvider(
+            method: "POST",
+            host: host,
+            path: path,
+            headers: nil,
+            queryParams: [ "user_id": userId ]
+        )
+    }
+}
+
+/// CachingTokenProvider delegates the fetching of tokens to an underlying `delegateProvider`, retrying on failure
+public class CachingTokenProvider: TokenProvider {
+    private let delegate: TokenProvider
+    private var token: String?
+
+    public init(delegateProvider: TokenProvider) {
+        self.delegate = delegateProvider
+    }
+
+    public func fetchToken(completionHandler: @escaping (TokenProviderResult) -> Void) {
+        // TODO: Implement:
+        // if token is nil or expired, delegate
+        // if call to delegate is in progress, attach to its completion handler
+        // return token
+
+        self.delegate.fetchToken(completionHandler: completionHandler)
+    }
+}
+
+/// RetryingTokenProvider delegates the fetching of tokens to an underlying `delegateProvider`, maintains a reference to the last fetched token and ensures
+/// that the delegate is only called when necessary:
+public class RetryingTokenProvider: TokenProvider {
+    public init(delegateProvider: TokenProvider, retryAttempts: UInt, retryDelayMs: UInt) {}
+
+    public func fetchToken(completionHandler: @escaping (TokenProviderResult) -> Void) {
+        // TODO: Implement:
+        // fetch token from delegate, with retries on failure
+    }
+}
+
+/// HTTPCallTokenProvider makes calls to a specified HTTP endpoint and expects to receive a token from it.
+///
+/// This class will make an HTTP request for every call to `fetchToken`, and in almost all cases it should be wrapped with caching and retry logic. See `httpTokenProvider(...)` for a factory method which returns a wrapped instance of this class.
+///
+/// - Parameters:
+///     - method: The HTTP method to use in the call, e.g. `GET`, `POST`
+///     - host: The host name to be called, e.g. `example.com`
+///     - path: The path component of the URL to be called, e.g. `/tokens`
+///     - headers: An optional map of headers to include in the request. Here you can supply the session or other credientials which your endpoint might require to authenticate the request.
+///     - queryParams: An optional map of query parameters to include in the request URL. Here you can supply the session or other credientials which your endpoint might require to authenticate the request.
```

Comment:
Wouldn't it be better to just just suggest putting session id as a header? I know it's not about URLs for a browser, or an end user (and they won't be shared) but putting such info as query parameters is generally considered not a good practice. For example within the backend infrastructure there may be logging software which will be logging URLs.

Reply:
Header is definitely the better solution, but this is for making a request to an endpoint the customer is providing. We don't know how their auth works, or whether it's any good, but we can't really dictate requirements on it.

Reply:
I understand I have just suggested not suggesting it in the docs for query params ;)